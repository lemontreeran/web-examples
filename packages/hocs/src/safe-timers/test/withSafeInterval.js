import React from 'react'
import { mount } from 'enzyme'

const Target = () => null

describe('withSafeInterval', () => {
  let withSafeInterval = null

  beforeAll(() => {
    jest.spyOn(global, 'setInterval').mockImplementation(() => 'id')
    jest.spyOn(global, 'clearInterval')
    jest.resetModules()

    withSafeInterval = require('../src').withSafeInterval
  })

  afterEach(() => {
    global.setInterval.mockClear()
    global.clearInterval.mockClear()
  })

  afterAll(() => {
    global.setInterval.mockRestore()
    global.clearInterval.mockRestore()
  })

  it('should pass props through', () => {
    const EnhancedTarget = withSafeInterval(Target)
    const wrapper = mount(
      <EnhancedTarget a={1} b={2} />
    )

    expect(wrapper.find(Target)).toMatchSnapshot()
  })

  it('should provide `setSafeInterval` prop and unsubscriber as its call return', () => {
    const callback = () => {}
    const EnhancedTarget = withSafeInterval(Target)
    const wrapper = mount(
      <EnhancedTarget />
    )
    const clearSafeInterval = wrapper.find(Target).prop('setSafeInterval')(callback, 'a', 'b')

    expect(global.setInterval.mock.calls).toMatchSnapshot()

    clearSafeInterval()

    expect(global.clearInterval.mock.calls).toMatchSnapshot()
  })

  it('should clear all safe intervals on unmount', () => {
    const EnhancedTarget = withSafeInterval(Target)
    const wrapper = mount(
      <EnhancedTarget />
    )
    const setSafeInterval = wrapper.find(Target).prop('setSafeInterval')

    setSafeInterval()
    setSafeInterval()
    setSafeInterval()

    wrapper.unmount()

    expect(global.clearInterval.mock.calls).toMatchSnapshot()
  })

  describe('display name', () => {
    const origNodeEnv = process.env.NODE_ENV

    afterAll(() => {
      process.env.NODE_ENV = origNodeEnv
    })

    it('should wrap display name in non-production env', () => {
      process.env.NODE_ENV = 'test'

      const EnhancedTarget = withSafeInterval(Target)
      const wrapper = mount(
        <EnhancedTarget />
      )

      expect(wrapper).toMatchSnapshot()
    })

    it('should not wrap display name in production env', () => {
      process.env.NODE_ENV = 'production'

      const EnhancedTarget = withSafeInterval(Target)
      const wrapper = mount(
        <EnhancedTarget />
      )

      expect(wrapper).toMatchSnapshot()
    })
  })
})
