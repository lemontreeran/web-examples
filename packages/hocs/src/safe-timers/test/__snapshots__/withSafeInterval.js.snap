// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withSafeInterval display name should not wrap display name in production env 1`] = `
<SafeTimer>
  <Target
    setSafeInterval={[Function]}
  />
</SafeTimer>
`;

exports[`withSafeInterval display name should wrap display name in non-production env 1`] = `
<withSafeInterval(Target)>
  <Target
    setSafeInterval={[Function]}
  />
</withSafeInterval(Target)>
`;

exports[`withSafeInterval should clear all safe intervals on unmount 1`] = `
Array [
  Array [
    "id",
  ],
  Array [
    "id",
  ],
  Array [
    "id",
  ],
]
`;

exports[`withSafeInterval should pass props through 1`] = `
<Target
  a={1}
  b={2}
  setSafeInterval={[Function]}
/>
`;

exports[`withSafeInterval should provide \`setSafeInterval\` prop and unsubscriber as its call return 1`] = `
Array [
  Array [
    [Function],
    "a",
    "b",
  ],
]
`;

exports[`withSafeInterval should provide \`setSafeInterval\` prop and unsubscriber as its call return 2`] = `
Array [
  Array [
    "id",
  ],
]
`;
